File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    answer = 0
    visited = [False] * n
    for hop in range(n) :
        if visited[hop] == False :
            BFS(n, computers, hop, visited)
           # DFS(n, computers, hop, visited)
            answer += 1
    return answer

def DFS(n, computers, hop, visited):
    visited[hop] = True 
    for i in range(n): 
        if i != hop and computers[i][hop] == 1 and visited[i] == False:    
                DFS(n, computers, i, visited)

def BFS(n, computers, hop, visited):
    queue = deque()
    queue.append(hop)
    while queue:
        hop = queue.pop()
        visited[hop] = True
        for i in range(n) :
            if i != hop and computers[i][hop] == 1 and visited[i] == False:   
                queue.append(i)

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.Queue;
import java.util.LinkedList;
class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        int[] visited = new int[n];
        
        for (Integer hop = 0; hop < n; hop++) {
            if (visited[hop] == 0) {
                visited = BFS(n, visited, hop, computers);
                answer += 1;
            }
        }
        
        return answer;
    }
    
    private int[] DFS(int n, int[] visited, int hop, int[][] computers) {
        visited[hop] = 1;
        for (Integer com = 0; com < n; com++) {
            if (com != hop && computers[com][hop] == 1 && visited[com] == 0) {
                visited =  DFS(n, visited, com, computers);
            }
        }
        
        return visited;
    }
    
    private int[] BFS(int n, int[] visited, int hop, int[][] computers) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(hop);
        
        while(!queue.isEmpty()) {
            Integer a = queue.poll();
            visited[a] = 1;
            for (Integer com = 0; com < n; com++) {
                if (com != a && computers[com][a] == 1 && visited[com] == 0) {
                    visited =  BFS(n, visited, com, computers);
                }
            }
        }
        return visited;
    }
}